### PR TITLE
Adapt to change in behavior of `pandas.Timestamp` constructor

### LINKED ISSUE #10932: ⚠️ Nightly upstream-dev CI failed ⚠️
[Workflow Run URL](https://github.com/pydata/xarray/actions/runs/19750488289)
<details><summary>Python 3.12 Test Summary</summary>

```
xarray/tests/test_coding_times.py::test_decode_cf_datetime_overflow[s]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[s]')
 +  where dtype('<M8[us]') = array('1677-09-20T00:00:00.000000', dtype='datetime64[us]').dtype
 +  and   dtype('<M8[s]') = <class 'numpy.dtype'>('=M8[s]')
 +    where <class 'numpy.dtype'> = np.dtype
xarray/tests/test_coding_times.py::test_decode_cf_datetime_overflow[ms]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[ms]')
 +  where dtype('<M8[us]') = array('1677-09-20T00:00:00.000000', dtype='datetime64[us]').dtype
 +  and   dtype('<M8[ms]') = <class 'numpy.dtype'>('=M8[ms]')
 +    where <class 'numpy.dtype'> = np.dtype
xarray/tests/test_coding_times.py::test_decode_standard_calendar_inside_timestamp_range[s-proleptic_gregorian]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[s]')
 +  where dtype('<M8[us]') = array(['2001-04-01T00:00:00.000000', '2001-04-01T01:00:00.000000',\n       '2001-04-01T02:00:00.000000', '2001-04-01T03...-30T21:00:00.000000',\n       '2001-04-30T22:00:00.000000', '2001-04-30T23:00:00.000000'],\n      dtype='datetime64[us]').dtype
 +  and   dtype('<M8[s]') = <class 'numpy.dtype'>('=M8[s]')
 +    where <class 'numpy.dtype'> = np.dtype
xarray/tests/test_coding_times.py::test_decode_standard_calendar_inside_timestamp_range[ms-proleptic_gregorian]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[ms]')
 +  where dtype('<M8[us]') = array(['2001-04-01T00:00:00.000000', '2001-04-01T01:00:00.000000',\n       '2001-04-01T02:00:00.000000', '2001-04-01T03...-30T21:00:00.000000',\n       '2001-04-30T22:00:00.000000', '2001-04-30T23:00:00.000000'],\n      dtype='datetime64[us]').dtype
 +  and   dtype('<M8[ms]') = <class 'numpy.dtype'>('=M8[ms]')
 +    where <class 'numpy.dtype'> = np.dtype
xarray/tests/test_coding_times.py::test_decode_standard_calendar_single_element_inside_timestamp_range[s-735368-proleptic_gregorian]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[s]')
 +  where dtype('<M8[us]') = array('2014-05-16T00:00:00.000000', dtype='datetime64[us]').dtype
 +  and   dtype('<M8[s]') = <class 'numpy.dtype'>('=M8[s]')
 +    where <class 'numpy.dtype'> = np.dtype
xarray/tests/test_coding_times.py::test_decode_standard_calendar_single_element_inside_timestamp_range[s-num_time1-proleptic_gregorian]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[s]')
 +  where dtype('<M8[us]') = array(['2014-05-16T00:00:00.000000'], dtype='datetime64[us]').dtype
 +  and   dtype('<M8[s]') = <class 'numpy.dtype'>('=M8[s]')
 +    where <class 'numpy.dtype'> = np.dtype
xarray/tests/test_coding_times.py::test_decode_standard_calendar_single_element_inside_timestamp_range[s-num_time2-proleptic_gregorian]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[s]')
 +  where dtype('<M8[us]') = array([['2014-05-16T00:00:00.000000']], dtype='datetime64[us]').dtype
 +  and   dtype('<M8[s]') = <class 'numpy.dtype'>('=M8[s]')
 +    where <class 'numpy.dtype'> = np.dtype
xarray/tests/test_coding_times.py::test_decode_standard_calendar_single_element_inside_timestamp_range[ms-735368-proleptic_gregorian]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[ms]')
 +  where dtype('<M8[us]') = array('2014-05-16T00:00:00.000000', dtype='datetime64[us]').dtype
 +  and   dtype('<M8[ms]') = <class 'numpy.dtype'>('=M8[ms]')
 +    where <class 'numpy.dtype'> = np.dtype
xarray/tests/test_coding_times.py::test_decode_standard_calendar_single_element_inside_timestamp_range[ms-num_time1-proleptic_gregorian]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[ms]')
 +  where dtype('<M8[us]') = array(['2014-05-16T00:00:00.000000'], dtype='datetime64[us]').dtype
 +  and   dtype('<M8[ms]') = <class 'numpy.dtype'>('=M8[ms]')
 +    where <class 'numpy.dtype'> = np.dtype
xarray/tests/test_coding_times.py::test_decode_standard_calendar_single_element_inside_timestamp_range[ms-num_time2-proleptic_gregorian]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[ms]')
 +  where dtype('<M8[us]') = array([['2014-05-16T00:00:00.000000']], dtype='datetime64[us]').dtype
 +  and   dtype('<M8[ms]') = <class 'numpy.dtype'>('=M8[ms]')
 +    where <class 'numpy.dtype'> = np.dtype
xarray/tests/test_coding_times.py::test_decode_standard_calendar_multidim_time_inside_timestamp_range[s-proleptic_gregorian]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[s]')
 +  where dtype('<M8[us]') = array([['2001-04-01T00:00:00.000000', '2001-05-01T00:00:00.000000'],\n       ['2001-04-02T00:00:00.000000', '2001-05-02...T00:00:00.000000'],\n       ['2001-04-05T00:00:00.000000', '2001-05-05T00:00:00.000000']],\n      dtype='datetime64[us]').dtype
 +  and   dtype('<M8[s]') = <class 'numpy.dtype'>('=M8[s]')
 +    where <class 'numpy.dtype'> = np.dtype
xarray/tests/test_coding_times.py::test_decode_standard_calendar_multidim_time_inside_timestamp_range[ms-proleptic_gregorian]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[ms]')
 +  where dtype('<M8[us]') = array([['2001-04-01T00:00:00.000000', '2001-05-01T00:00:00.000000'],\n       ['2001-04-02T00:00:00.000000', '2001-05-02...T00:00:00.000000'],\n       ['2001-04-05T00:00:00.000000', '2001-05-05T00:00:00.000000']],\n      dtype='datetime64[us]').dtype
 +  and   dtype('<M8[ms]') = <class 'numpy.dtype'>('=M8[ms]')
 +    where <class 'numpy.dtype'> = np.dtype
xarray/tests/test_coding_times.py::test_decode_multidim_time_outside_timestamp_range[s-proleptic_gregorian]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[s]')
 +  where dtype('<M8[us]') = array([['0001-04-01T00:00:00.000000', '0001-05-01T00:00:00.000000'],\n       ['0001-04-02T00:00:00.000000', '0001-05-02...T00:00:00.000000'],\n       ['0001-04-05T00:00:00.000000', '0001-05-05T00:00:00.000000']],\n      dtype='datetime64[us]').dtype
xarray/tests/test_coding_times.py::test_decode_multidim_time_outside_timestamp_range[ms-proleptic_gregorian]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[ms]')
 +  where dtype('<M8[us]') = array([['0001-04-01T00:00:00.000000', '0001-05-01T00:00:00.000000'],\n       ['0001-04-02T00:00:00.000000', '0001-05-02...T00:00:00.000000'],\n       ['0001-04-05T00:00:00.000000', '0001-05-05T00:00:00.000000']],\n      dtype='datetime64[us]').dtype
xarray/tests/test_coding_times.py::test_decoded_cf_datetime_array_2d[s]: AssertionError: assert dtype('<M8[us]') == 'datetime64[s]'
 +  where dtype('<M8[us]') = <xarray.Variable (x: 2, y: 2)> Size: 32B\n[4 values with dtype=datetime64[us]].dtype
xarray/tests/test_coding_times.py::test_decoded_cf_datetime_array_2d[ms]: AssertionError: assert dtype('<M8[us]') == 'datetime64[ms]'
 +  where dtype('<M8[us]') = <xarray.Variable (x: 2, y: 2)> Size: 32B\n[4 values with dtype=datetime64[us]].dtype
xarray/tests/test_coding_times.py::test_decode_cf[s-gregorian]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[s]')
 +  where dtype('<M8[us]') = <xarray.DataArray 'test' (time: 3)> Size: 24B\n[3 values with dtype=datetime64[us]]\nCoordinates:\n  * time     (time) datetime64[us] 24B 2001-01-02 2001-01-03 2001-01-04.dtype
 +    where <xarray.DataArray 'test' (time: 3)> Size: 24B\n[3 values with dtype=datetime64[us]]\nCoordinates:\n  * time     (time) datetime64[us] 24B 2001-01-02 2001-01-03 2001-01-04 = <xarray.Dataset> Size: 48B\nDimensions:  (time: 3)\nCoordinates:\n  * time     (time) datetime64[us] 24B 2001-01-02 2001-01-03 2001-01-04\nData variables:\n    test     (time) datetime64[us] 24B ....test
 +  and   dtype('<M8[s]') = <class 'numpy.dtype'>('=M8[s]')
 +    where <class 'numpy.dtype'> = np.dtype
xarray/tests/test_coding_times.py::test_decode_cf[s-proleptic_gregorian]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[s]')
 +  where dtype('<M8[us]') = <xarray.DataArray 'test' (time: 3)> Size: 24B\n[3 values with dtype=datetime64[us]]\nCoordinates:\n  * time     (time) datetime64[us] 24B 2001-01-02 2001-01-03 2001-01-04.dtype
 +    where <xarray.DataArray 'test' (time: 3)> Size: 24B\n[3 values with dtype=datetime64[us]]\nCoordinates:\n  * time     (time) datetime64[us] 24B 2001-01-02 2001-01-03 2001-01-04 = <xarray.Dataset> Size: 48B\nDimensions:  (time: 3)\nCoordinates:\n  * time     (time) datetime64[us] 24B 2001-01-02 2001-01-03 2001-01-04\nData variables:\n    test     (time) datetime64[us] 24B ....test
 +  and   dtype('<M8[s]') = <class 'numpy.dtype'>('=M8[s]')
 +    where <class 'numpy.dtype'> = np.dtype
xarray/tests/test_coding_times.py::test_decode_cf[s-standard]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[s]')
 +  where dtype('<M8[us]') = <xarray.DataArray 'test' (time: 3)> Size: 24B\n[3 values with dtype=datetime64[us]]\nCoordinates:\n  * time     (time) datetime64[us] 24B 2001-01-02 2001-01-03 2001-01-04.dtype
 +    where <xarray.DataArray 'test' (time: 3)> Size: 24B\n[3 values with dtype=datetime64[us]]\nCoordinates:\n  * time     (time) datetime64[us] 24B 2001-01-02 2001-01-03 2001-01-04 = <xarray.Dataset> Size: 48B\nDimensions:  (time: 3)\nCoordinates:\n  * time     (time) datetime64[us] 24B 2001-01-02 2001-01-03 2001-01-04\nData variables:\n    test     (time) datetime64[us] 24B ....test
 +  and   dtype('<M8[s]') = <class 'numpy.dtype'>('=M8[s]')
 +    where <class 'numpy.dtype'> = np.dtype
xarray/tests/test_coding_times.py::test_decode_cf[ms-gregorian]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[ms]')
 +  where dtype('<M8[us]') = <xarray.DataArray 'test' (time: 3)> Size: 24B\n[3 values with dtype=datetime64[us]]\nCoordinates:\n  * time     (time) datetime64[us] 24B 2001-01-02 2001-01-03 2001-01-04.dtype
 +    where <xarray.DataArray 'test' (time: 3)> Size: 24B\n[3 values with dtype=datetime64[us]]\nCoordinates:\n  * time     (time) datetime64[us] 24B 2001-01-02 2001-01-03 2001-01-04 = <xarray.Dataset> Size: 48B\nDimensions:  (time: 3)\nCoordinates:\n  * time     (time) datetime64[us] 24B 2001-01-02 2001-01-03 2001-01-04\nData variables:\n    test     (time) datetime64[us] 24B ....test
 +  and   dtype('<M8[ms]') = <class 'numpy.dtype'>('=M8[ms]')
 +    where <class 'numpy.dtype'> = np.dtype
xarray/tests/test_coding_times.py::test_decode_cf[ms-proleptic_gregorian]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[ms]')
 +  where dtype('<M8[us]') = <xarray.DataArray 'test' (time: 3)> Size: 24B\n[3 values with dtype=datetime64[us]]\nCoordinates:\n  * time     (time) datetime64[us] 24B 2001-01-02 2001-01-03 2001-01-04.dtype
 +    where <xarray.DataArray 'test' (time: 3)> Size: 24B\n[3 values with dtype=datetime64[us]]\nCoordinates:\n  * time     (time) datetime64[us] 24B 2001-01-02 2001-01-03 2001-01-04 = <xarray.Dataset> Size: 48B\nDimensions:  (time: 3)\nCoordinates:\n  * time     (time) datetime64[us] 24B 2001-01-02 2001-01-03 2001-01-04\nData variables:\n    test     (time) datetime64[us] 24B ....test
 +  and   dtype('<M8[ms]') = <class 'numpy.dtype'>('=M8[ms]')
 +    where <class 'numpy.dtype'> = np.dtype
xarray/tests/test_coding_times.py::test_decode_cf[ms-standard]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[ms]')
 +  where dtype('<M8[us]') = <xarray.DataArray 'test' (time: 3)> Size: 24B\n[3 values with dtype=datetime64[us]]\nCoordinates:\n  * time     (time) datetime64[us] 24B 2001-01-02 2001-01-03 2001-01-04.dtype
 +    where <xarray.DataArray 'test' (time: 3)> Size: 24B\n[3 values with dtype=datetime64[us]]\nCoordinates:\n  * time     (time) datetime64[us] 24B 2001-01-02 2001-01-03 2001-01-04 = <xarray.Dataset> Size: 48B\nDimensions:  (time: 3)\nCoordinates:\n  * time     (time) datetime64[us] 24B 2001-01-02 2001-01-03 2001-01-04\nData variables:\n    test     (time) datetime64[us] 24B ....test
 +  and   dtype('<M8[ms]') = <class 'numpy.dtype'>('=M8[ms]')
 +    where <class 'numpy.dtype'> = np.dtype
xarray/tests/test_coding_times.py::test_decode_cf_time_bounds[s]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[s]')
 +  where dtype('<M8[us]') = <xarray.DataArray 'time_bnds' (time: 3, nbnd: 2)> Size: 48B\n[6 values with dtype=datetime64[us]]\nCoordinates:\n  * time     (time) datetime64[us] 24B 2001-01-02 2001-01-03 2001-01-04\nDimensions without coordinates: nbnd.dtype
 +    where <xarray.DataArray 'time_bnds' (time: 3, nbnd: 2)> Size: 48B\n[6 values with dtype=datetime64[us]]\nCoordinates:\n  * time     (time) datetime64[us] 24B 2001-01-02 2001-01-03 2001-01-04\nDimensions without coordinates: nbnd = <xarray.Dataset> Size: 72B\nDimensions:    (time: 3, nbnd: 2)\nCoordinates:\n  * time       (time) datetime64[us] 24B 200...1-03 2001-01-04\nDimensions without coordinates: nbnd\nData variables:\n    time_bnds  (time, nbnd) datetime64[us] 48B ....time_bnds
 +  and   dtype('<M8[s]') = <class 'numpy.dtype'>('=M8[s]')
 +    where <class 'numpy.dtype'> = np.dtype
xarray/tests/test_coding_times.py::test_decode_cf_time_bounds[ms]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[ms]')
 +  where dtype('<M8[us]') = <xarray.DataArray 'time_bnds' (time: 3, nbnd: 2)> Size: 48B\n[6 values with dtype=datetime64[us]]\nCoordinates:\n  * time     (time) datetime64[us] 24B 2001-01-02 2001-01-03 2001-01-04\nDimensions without coordinates: nbnd.dtype
 +    where <xarray.DataArray 'time_bnds' (time: 3, nbnd: 2)> Size: 48B\n[6 values with dtype=datetime64[us]]\nCoordinates:\n  * time     (time) datetime64[us] 24B 2001-01-02 2001-01-03 2001-01-04\nDimensions without coordinates: nbnd = <xarray.Dataset> Size: 72B\nDimensions:    (time: 3, nbnd: 2)\nCoordinates:\n  * time       (time) datetime64[us] 24B 200...1-03 2001-01-04\nDimensions without coordinates: nbnd\nData variables:\n    time_bnds  (time, nbnd) datetime64[us] 48B ....time_bnds
 +  and   dtype('<M8[ms]') = <class 'numpy.dtype'>('=M8[ms]')
 +    where <class 'numpy.dtype'> = np.dtype
xarray/tests/test_coding_times.py::test_decode_float_datetime_with_decimals[ms]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[ms]')
 +  where dtype('<M8[us]') = array(['2000-01-01T00:00:00.000000', '2000-01-01T00:00:00.125000',\n       '2000-01-01T00:00:00.250000', '2000-01-01T00:00:00.375000',\n       '2000-01-01T00:00:00.750000', '2000-01-01T00:00:01.000000'],\n      dtype='datetime64[us]').dtype
 +  and   dtype('<M8[ms]') = array(['2000-01-01T00:00:00.000', '2000-01-01T00:00:00.125',\n       '2000-01-01T00:00:00.250', '2000-01-01T00:00:00.375',\n       '2000-01-01T00:00:00.750', '2000-01-01T00:00:01.000'],\n      dtype='datetime64[ms]').dtype
xarray/tests/test_coding_times.py::test_coding_float_datetime_warning[s-0.123]: Failed: DID NOT WARN. No warnings of type (<class 'xarray.coding.common.SerializationWarning'>,) matching the regex were emitted.
 Regex: Can't decode floating point datetimes to 's'
 Emitted warnings: [ SerializationWarning("Can't decode floating point datetimes to 'us' without precision loss; decoding to 'ns' instead. To silence this warning pass decode_times=CFDatetimeCoder(time_unit='ns') to your opening function.")].
xarray/tests/test_coding_times.py::test_coding_float_datetime_warning[ms-0.1234]: Failed: DID NOT WARN. No warnings of type (<class 'xarray.coding.common.SerializationWarning'>,) matching the regex were emitted.
 Regex: Can't decode floating point datetimes to 'ms'
 Emitted warnings: [ SerializationWarning("Can't decode floating point datetimes to 'us' without precision loss; decoding to 'ns' instead. To silence this warning pass decode_times=CFDatetimeCoder(time_unit='ns') to your opening function.")].
xarray/tests/test_coding_times.py::test_encode_cf_datetime_datetime64_via_dask[s-pandas-encoding-with-prescribed-units-and-dtype]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[s]')
 +  where dtype('<M8[us]') = array(['1700-01-01T00:00:00.000000', '1700-01-02T00:00:00.000000',\n       '1700-01-03T00:00:00.000000'], dtype='datetime64[us]').dtype
 +  and   dtype('<M8[s]') = dask.array<array, shape=(3,), dtype=datetime64[s], chunksize=(1,), chunktype=pandas.DatetimeIndex>.dtype
xarray/tests/test_coding_times.py::test_encode_cf_datetime_datetime64_via_dask[s-mixed-cftime-pandas-encoding-with-prescribed-units-and-dtype]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[s]')
 +  where dtype('<M8[us]') = array(['1700-01-01T00:00:00.000000', '1950-01-01T00:00:00.000000',\n       '2200-01-01T00:00:00.000000'], dtype='datetime64[us]').dtype
 +  and   dtype('<M8[s]') = dask.array<array, shape=(3,), dtype=datetime64[s], chunksize=(1,), chunktype=pandas.DatetimeIndex>.dtype
xarray/tests/test_coding_times.py::test_encode_cf_datetime_datetime64_via_dask[s-pandas-encoding-with-default-units-and-dtype]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[s]')
 +  where dtype('<M8[us]') = array(['1700-01-01T00:00:00.000000', '1950-01-01T00:00:00.000000',\n       '2200-01-01T00:00:00.000000'], dtype='datetime64[us]').dtype
 +  and   dtype('<M8[s]') = dask.array<array, shape=(3,), dtype=datetime64[s], chunksize=(1,), chunktype=pandas.DatetimeIndex>.dtype
xarray/tests/test_coding_times.py::test_encode_cf_datetime_datetime64_via_dask[ms-pandas-encoding-with-prescribed-units-and-dtype]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[ms]')
 +  where dtype('<M8[us]') = array(['1700-01-01T00:00:00.000000', '1700-01-02T00:00:00.000000',\n       '1700-01-03T00:00:00.000000'], dtype='datetime64[us]').dtype
 +  and   dtype('<M8[ms]') = dask.array<array, shape=(3,), dtype=datetime64[ms], chunksize=(1,), chunktype=pandas.DatetimeIndex>.dtype
xarray/tests/test_coding_times.py::test_encode_cf_datetime_datetime64_via_dask[ms-mixed-cftime-pandas-encoding-with-prescribed-units-and-dtype]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[ms]')
 +  where dtype('<M8[us]') = array(['1700-01-01T00:00:00.000000', '1950-01-01T00:00:00.000000',\n       '2200-01-01T00:00:00.000000'], dtype='datetime64[us]').dtype
 +  and   dtype('<M8[ms]') = dask.array<array, shape=(3,), dtype=datetime64[ms], chunksize=(1,), chunktype=pandas.DatetimeIndex>.dtype
xarray/tests/test_coding_times.py::test_encode_cf_datetime_datetime64_via_dask[ms-pandas-encoding-with-default-units-and-dtype]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[ms]')
 +  where dtype('<M8[us]') = array(['1700-01-01T00:00:00.000000', '1950-01-01T00:00:00.000000',\n       '2200-01-01T00:00:00.000000'], dtype='datetime64[us]').dtype
 +  and   dtype('<M8[ms]') = dask.array<array, shape=(3,), dtype=datetime64[ms], chunksize=(1,), chunktype=pandas.DatetimeIndex>.dtype
xarray/tests/test_coding_times.py::test_roundtrip_empty_datetime64_array[s]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[s]')
 +  where dtype('<M8[us]') = <xarray.Variable (time: 0)> Size: 0B\n[0 values with dtype=datetime64[us]].dtype
 +  and   dtype('<M8[s]') = <xarray.Variable (time: 0)> Size: 0B\narray([], dtype='datetime64[s]').dtype
xarray/tests/test_coding_times.py::test_roundtrip_empty_datetime64_array[ms]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[ms]')
 +  where dtype('<M8[us]') = <xarray.Variable (time: 0)> Size: 0B\n[0 values with dtype=datetime64[us]].dtype
 +  and   dtype('<M8[ms]') = <xarray.Variable (time: 0)> Size: 0B\narray([], dtype='datetime64[ms]').dtype
xarray/tests/test_conventions.py::TestDecodeCF::test_dataset_repr_with_netcdf4_datetimes[s]: AssertionError: assert '(time) datetime64[s]' in '<xarray.Dataset> Size: 16B\nDimensions:  (time: 2)\nCoordinates:\n  * time     (time) datetime64[us] 16B 1900-01-01 1900-01-02\nData variables:\n    *empty*'
 +  where '<xarray.Dataset> Size: 16B\nDimensions:  (time: 2)\nCoordinates:\n  * time     (time) datetime64[us] 16B 1900-01-01 1900-01-02\nData variables:\n    *empty*' = repr(<xarray.Dataset> Size: 16B\nDimensions:  (time: 2)\nCoordinates:\n  * time     (time) datetime64[us] 16B 1900-01-01 1900-01-02\nData variables:\n    *empty*)
xarray/tests/test_conventions.py::TestDecodeCF::test_dataset_repr_with_netcdf4_datetimes[ms]: AssertionError: assert '(time) datetime64[ms]' in '<xarray.Dataset> Size: 16B\nDimensions:  (time: 2)\nCoordinates:\n  * time     (time) datetime64[us] 16B 1900-01-01 1900-01-02\nData variables:\n    *empty*'
 +  where '<xarray.Dataset> Size: 16B\nDimensions:  (time: 2)\nCoordinates:\n  * time     (time) datetime64[us] 16B 1900-01-01 1900-01-02\nData variables:\n    *empty*' = repr(<xarray.Dataset> Size: 16B\nDimensions:  (time: 2)\nCoordinates:\n  * time     (time) datetime64[us] 16B 1900-01-01 1900-01-02\nData variables:\n    *empty*)
xarray/tests/test_conventions.py::TestDecodeCF::test_decode_cf_time_kwargs[s]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[s]')
 +  where dtype('<M8[us]') = <xarray.DataArray 'time' (time: 3)> Size: 24B\narray(['2000-01-02T00:00:00.000000', '2000-01-03T00:00:00.000000',\n     ....000000'], dtype='datetime64[us]')\nCoordinates:\n  * time     (time) datetime64[us] 24B 2000-01-02 2000-01-03 2000-01-04.dtype
 +    where <xarray.DataArray 'time' (time: 3)> Size: 24B\narray(['2000-01-02T00:00:00.000000', '2000-01-03T00:00:00.000000',\n     ....000000'], dtype='datetime64[us]')\nCoordinates:\n  * time     (time) datetime64[us] 24B 2000-01-02 2000-01-03 2000-01-04 = <xarray.Dataset> Size: 120B\nDimensions:    (timedelta: 3, time: 3)\nCoordinates:\n  * timedelta  (timedelta) timedelta64... 2000-01-03 2000-01-04\nData variables:\n    a          (time, timedelta) float64 72B 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0.time
 +  and   dtype('<M8[s]') = <class 'numpy.dtype'>('M8[s]')
 +    where <class 'numpy.dtype'> = np.dtype
xarray/tests/test_conventions.py::TestDecodeCF::test_decode_cf_time_kwargs[ms]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[ms]')
 +  where dtype('<M8[us]') = <xarray.DataArray 'time' (time: 3)> Size: 24B\narray(['2000-01-02T00:00:00.000000', '2000-01-03T00:00:00.000000',\n     ....000000'], dtype='datetime64[us]')\nCoordinates:\n  * time     (time) datetime64[us] 24B 2000-01-02 2000-01-03 2000-01-04.dtype
 +    where <xarray.DataArray 'time' (time: 3)> Size: 24B\narray(['2000-01-02T00:00:00.000000', '2000-01-03T00:00:00.000000',\n     ....000000'], dtype='datetime64[us]')\nCoordinates:\n  * time     (time) datetime64[us] 24B 2000-01-02 2000-01-03 2000-01-04 = <xarray.Dataset> Size: 120B\nDimensions:    (timedelta: 3, time: 3)\nCoordinates:\n  * timedelta  (timedelta) timedelta64... 2000-01-03 2000-01-04\nData variables:\n    a          (time, timedelta) float64 72B 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0.time
 +  and   dtype('<M8[ms]') = <class 'numpy.dtype'>('M8[ms]')
 +    where <class 'numpy.dtype'> = np.dtype
xarray/tests/test_dataarray.py::TestDataArray::test_curvefit_helpers: Failed: DID NOT RAISE <class 'ValueError'>
xarray/tests/test_variable.py::TestVariable::test_datetime64_conversion_scalar[values1-s]: AssertionError: assert dtype('<M8[us]') == dtype('<M8[s]')
 +  where dtype('<M8[us]') = <xarray.Variable ()> Size: 8B\narray('2000-01-01T00:00:00.000000', dtype='datetime64[us]').dtype
 +  and   dtype('<M8[s]') = <class 'numpy.dtype'>('datetime64[s]')
 +    where <class 'numpy.dtype'> = np.dtype
xarray/tests/test_variable.py::TestVariable::test_0d_datetime: AssertionError: assert dtype('<M8[us]') == dtype('<M8[s]')
 +  where dtype('<M8[us]') = <xarray.Variable ()> Size: 8B\narray('2000-01-01T00:00:00.000000', dtype='datetime64[us]').dtype
 +  and   dtype('<M8[s]') = <class 'numpy.dtype'>('datetime64[s]')
 +    where <class 'numpy.dtype'> = np.dtype
```

</details>

### PR DESCRIPTION
**Pull request by spencerkclark**
This PR resolves the upstream test failures indicated in #10932 by introducing a function to automatically cast a `pandas.Timestamp` to the coarsest resolution without changing its value, which we depend on in decoding the reference date of datetime data.  Previously the `Timestamp` constructor would approximately accomplish this for us, but this was changed in pandas-dev/pandas#62801.

In some cases the behavior of this function will be different than the previous `Timestamp` constructor, since this function does not parse the string for this information.  For example, previously the `Timestamp` constructor would assume that millisecond resolution was needed to represent this time:

```
>>> pd.Timestamp("2000-01-01T00:00:00.00").unit
'ms'
```

but strictly speaking only `"s"` resolution is needed, which is what this new approach will choose. In a certain way, however, this may ultimately be more consistent with [how we infer the resolution of the data values themselves](https://github.com/pydata/xarray/blob/a8efe17e0922fc3e69d08a1e3669e8371fb41843/xarray/coding/times.py#L440-L451). I did not add a what's new entry since this is somewhat of an internal change, but I can add one if we feel the change in behavior of this edge case is worth noting.

- [x] Closes #10932

cc: @kmuehlbauer

### CODE REVIEW

File: xarray/tests/test_variable.py
Comment (spencerkclark) [inline xarray/tests/test_variable.py:1]: The changes here are unrelated to the change in time decoding—here we were merely relying on the previous `Timestamp` constructor behavior in setting up the tests.

File: xarray/coding/times.py [hunk 1]
```diff
@@ -293,6 +295,21 @@ def _maybe_strip_tz_from_timestamp(date: pd.Timestamp) -> pd.Timestamp:
...
+    # constructor used to automatically infer this from the resolution of the
+    # string, but this behavior was changed in pandas-dev/pandas#62801. This
+    # function allows us to approximately restore the old behavior in a way
+    # that is perhaps more consistent with how we infer the resolution of the
+    # data values themselves.
+    for unit in _ORDERED_PANDAS_TIME_RESOLUTIONS:
+        coarsest_timestamp = timestamp.as_unit(unit)
+        if coarsest_timestamp == timestamp:
+            break
+    return coarsest_timestamp
```
Comment (shoyer) [inline xarray/coding/times.py]: This looks a little safer to me (recall that even finer resolution datetime64 dtypes than `ns` are possible, albeit rarely seen in practice):
```suggestion
    for unit in _ORDERED_PANDAS_TIME_RESOLUTIONS:
        coarsest_timestamp = timestamp.as_unit(unit)
        if coarsest_timestamp == timestamp:
            return coarsest_timestamp
    return timestamp
```

Comment (spencerkclark) [reply xarray/coding/times.py]: Thanks, indeed no reason not to use this safer approach, even if `Timestamp` objects  (unlike `datetime64` values) do not currently support resolutions finer than `"ns"`.

**Comment by spencerkclark** [comment]
Thanks @shoyer and @kmuehlbauer
